### PR TITLE
Fix 6394 -- DetailsList columnReorderOptions causing scrolling

### LINKED ISSUE #6394: DetailsList: Horizontal and vertical scrolling when items is empty and column reorder is enabled
### Please provide a reproduction of the bug in a codepen:
 
https://codepen.io/goodoldneon/pen/dqgeGE

### Bug Report 
 
- __Package version(s)__: 6.70.0
- __Browser and OS versions__: Chrome 69.0.3497.92 and macOS 10.13.6
 
#### Priorities and help requested:
 
Are you willing to submit a PR to fix? I'll try
 
Requested priority: Normal
 
Products/sites affected: N/A
 
#### Describe the issue:
 
The `DetailsList` component scrolls horizontally and vertically when both:
- `props.items === []`
- `props.columnReorderOptions !== null`

The `DetailsList` component scrolls horizontally when:
- `props.columnReorderOptions !== null`
 
#### Expected behavior:
 
No scrolling inside `DetailsList`.

### PR DESCRIPTION
**Pull request by goodoldneon**
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #6394
- [ ] Include a change request file using `$ npm run change`

#### Description of changes

Enabling `columnReorderOptions` in the `DetailsList` causes "drop hints" to appear. Due the drop hints' CSS stying, the `DetailsList` component undesirably has scrolling.

Giving the drop hint container `overflow: hidden` fixes the scrolling issue, but prevents the drop hints from appearing. Since drop hints only need to be visible while the user is dragging a column header, this PR conditionally sets `overflow: visible` while the header is being dragged.

While this PR fixes the issue, I don't know if it's the "right" method. To be honest, this fix feels a little hacky.

Additionally, this PR is not ready to merge since it fails a test (`src/components/DetailsList/DetailsHeader.test.tsx`). This is because the test expects the drop hints to be "visible", even when they aren't actually visible (when the user is not dragging a headers):

```js
expect(dropHintElementChildren.item(0).getAttribute('style')).toEqual('visibility: visible;');
```

#### Focus areas to test

This can be seen in the `DetailsList` "DragDrop" example. When column reordering is:
- Disabled, then the component does not scroll horizontally.
- Enabled, then the component scrolls horizontally.

### CODE REVIEW

**Comment by KevinTCoughlin** [comment]
cc: @arkogupta

**Comment by arkogupta** [comment]
Hi @goodoldneon ,

After checking the demo app using your branch, it seems there are still some issues. Yes the scrolling is not present on enabling column reorder but after we try to drag a header the scroll bar appears and it stays. Ideally, it should appear only if try to drag outside the viewport to view the last drop hint completely. Otherwise it shouldn't be present.

**Comment by goodoldneon** [comment]
You're right, @arkogupta. I didn't notice that since scrollbars are only visible on a Mac while scrolling.

I'll give this issue another shot, but I think it might be beyond my skill level.

**Comment by Vitalius1** [comment]
@goodoldneon @arkogupta 
Can you kindly update on the status of this PR? Closing or merging? Thanks!

**Comment by goodoldneon** [comment]
@Vitalius1 

The issue seems to be fixed when I change `div.ms-DetailsList` to be `overflow: visible`. That div is currently `overflow-x: auto; overflow-y: visible;`. What's the reason for `overflow-x: auto`?

**Comment by KevinTCoughlin** [comment]
This change fixes the "at rest" state regarding horizontal scrolling, but it still occurs during drag and drop as pointed out in this PR description. @arkogupta @laxmikankanala are you able to help address this style regression due to drag and drop hints? Tagging you folks as I believe you implemented the behavior and have more context than I.

**Comment by KevinTCoughlin** [comment]
Here's a GIF capturing the horizontal scrolling behavior we're trying to fix:

detailslist_dnd_regr

The drag & drop hints are taking up horizontal space when visible causing the contents to shift larger than the containing element which leads to horizontal overflow. It seems like we'd need to either position them in a way that they don't take up horizontal space within the container or use pseudo elements to do the same.

**Comment by micahgodbolt** [comment]
@goodoldneon Are you still waiting for some information? I'm happy to help get this PR moving again if it's still a valuable addition.

**Comment by goodoldneon** [comment]
> @goodoldneon Are you still waiting for some information? I'm happy to help get this PR moving again if it's still a valuable addition.

Yea, I'm wondering why `div.ms-DetailsList` has the CSS property `overflow-x: auto`. The scrolling issue goes away when I give it the CSS property `overflow: visible` (though I'm not sure if there are any side-effects).

**Comment by cliffkoh** [comment]
Pinged @micahgodbolt and @KevinTCoughlin

**Comment by micahgodbolt** [comment]
@goodoldneon sorry about the delay here. I believe there are situations where the details list is allowed to scroll. In this case it appears that horizontal scrolling should never happen, so we could probably make that style conditional.

**Comment by micahgodbolt** [comment]
Tests are currently failing 
image

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.styles.ts [hunk 1]
```diff
@@ -339,7 +339,8 @@ export const getStyles = (props: IDetailsHeaderStyleProps): IDetailsHeaderStyles
     dropHintStyle: [
       {
         display: 'inline-block',
-        position: 'absolute'
+        position: 'absolute',
+        overflow: 'hidden'
```
Comment (micahgodbolt) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.styles.ts]: Can you quickly explain why there are 2 different overflow changes in this PR? This one in styles, and the other in resetDropHints. Are those the same css values? Or two different elements?

Comment (goodoldneon) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.styles.ts]: I was taking the wrong approach when I submitted the PR. The issue seems to be resolved when I just change `.ms-DetailsList` to `overflow: visible`. That class already has `overflow-y: visible`, so I was wondering why it had `overflow-x: auto`.

The `DetailsList` component doesn't seem to support x-scrolling, even when it's so narrow that some columns aren't visible. So I don't see the harm in making `.ms-DetailsList` have `overflow: visible`.

If you agree with the the `overflow: visible` solution, I can submit another PR.

Comment (arkogupta) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.styles.ts]: But making `overflow-x: visible` would mean even when content starts overflowing horizontally, we wouldn’t have a scroll bar.

image

Comment (goodoldneon) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.styles.ts]: The scrollbar doesn't do anything right now. When content overflows horizontally, columns are hidden and/or cell contents get ellipses. Horizontal scrolling doesn't do anything besides create an unnecessary horizontal scrollbar.

Comment (dzearing) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.styles.ts]: There is a mode in DetailsList to allow for horizontal scrolling. See the `constrainMode` prop.

Comment (goodoldneon) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.styles.ts]: I took a look at `constrainMode`. It defaults to `horizontalConstrained`, which is setting `overflowX: 'auto'`. Is the purpose of this mode to allow horizontal scrolling when there isn't enough width to display all columns? If so, then it isn't working as expected. The issue raised by this PR is caused by `horizontalConstrained`, and I'm not seeing any benefit from `horizontalConstrained`.

Setting `constrainMode={unconstrained}` solves the issue raised by this PR.

Please let me know what `constrainMode={horizontalConstrained}` is supposed to do.

**Comment by KevinTCoughlin** [comment]
> 
> 
> This pull request has been automatically marked as stale because it has marked as requiring author feedback but has not had any activity for **7 days**. It will be closed if no further activity occurs **within 7 days of this comment**. Thank you for your contributions to Fabric React!

@cliffkoh no _rush_, but is the above label removal intentional by Fabric bot?

**Comment by cliffkoh** [comment]
Yes the author responded in a comment thread.

**Comment by cliffkoh** [comment]
@dzearing could you reply to @goodoldneon please?

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.styles.ts (hunk 1 above)
Comment (KevinTCoughlin) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.styles.ts]: Judging from the _only_ usage of that prop, it seems like the intent of `horizontalConstrained` is to ensure that the component overflows in the `X` direction by setting the according styles:

https://github.com/OfficeDev/office-ui-fabric-react/blob/ea8aacd189d53a164e95c3535c3170e447314ec2/packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.styles.ts#L49

I don't have full context as the prop predates my involvement with OUFR, but that is my reading of it.

Comment (goodoldneon) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.styles.ts]: Yea, I agree that seems like the purpose, but does it actually accomplish that? The only effect it seems to have is creating a scrollbar that moves a few pixels (i.e. the issue which prompted this PR).

**Comment by dzearing** [comment]
I will pull the changes down tomorrow to debug more thoroughly to assess. Sorry for the delay.

**Comment by goodoldneon** [comment]
> I will pull the changes down tomorrow to debug more thoroughly to assess. Sorry for the delay.

The changes proposed in this PR are irrelevant, since `constrainMode={0}` solves the issue. However, I think it's unintuitive that `constrainMode={1}` is the default, even though it doesn't seem to have any benefit. Either there's something I'm not seeing, or `constrainMode={1}` isn't serving a purpose.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.styles.ts (hunk 1 above)
Comment (KevinTCoughlin) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.styles.ts]: I think there's still a valid use-case for having a prop which constrains width and therefore requires `overflow-x: auto`. I've isolated the root cause of the above linked issue. Let me solve that and then perhaps I can have time to demonstrate the `isHorizontalConstrained` prop's usefulness. If it isn't useful then we can deprecate.

**Comment by KevinTCoughlin** [comment]
> > I will pull the changes down tomorrow to debug more thoroughly to assess. Sorry for the delay.
> 
> The changes proposed in this PR are irrelevant, since `constrainMode={0}` solves the issue. However, I think it's unintuitive that `constrainMode={1}` is the default, even though it doesn't seem to have any benefit. Either there's something I'm not seeing, or `constrainMode={1}` isn't serving a purpose.

@goodoldneon, I did some investigation and there _does indeed_ seem to be some funkiness going on. You can see in this [Codepen](https://codepen.io/kevintcoughlin/pen/NEoPqE) that I am not enabling `ConstrainMode.horizontalConstrained`, yet the `.is-horizontalConstrained` className is _still applied_. I've opened https://github.com/OfficeDev/office-ui-fabric-react/issues/7279 for discussion and additional findings.

Regarding this PR, I'm going to close this as we have a team looking into the root cause as described here https://github.com/OfficeDev/office-ui-fabric-react/issues/6394#issuecomment-443364219 and this will not solve that as-is.

I hope you find the above agreeable. If not feel free to reach out to me in the above linked issue and we'll get it sorted out.

Thank you for bringing this to our attention.

**Pull request closed without merging**